### PR TITLE
Comment on a ticket

### PR DESCRIPTION
**Pull request by james2509**
I have added Comment on a ticket functionality, as follows:

- Implement a comment structure that validates that a comment is not an empty string. I see this structure being able to be extended to include other useful properties. For example, CreateDate/UpdateDate (order comments on a ticket), CreatedBy/UpdatedBy (keep track of who the comment belongs to) and DeletedFlag if we wanted to allow logical deletion of comments.
- I updated the ticket structure to include a new comments property which is a collection of comment structures. After doing a bit of digging on collections I went for a vector as felt that works for this scenario.
- I updated TicketStore to include a function to add a comment to a ticket. I thought about treating comments as a completely separate entity and having a CommentStore and store the ticket id against a comment. I ended up going the route I did as felt the comments are still a property of the ticket so in code it felt right they would be managed via the same store and available on the ticket. In the persistence layer we may want to store them separately and join between the two.

### CODE REVIEW

File: src/models/comment.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
+use std::error::Error;
+use std::fmt;
+
+#[derive(PartialEq, Debug, Clone, Hash, Eq)]
+/// Represents a comment on a [Ticket](Ticket)
+/// Wraps a string and checks that it is not empty when set
+pub struct Comment {
+    pub comment: String,
```
Comment (LukeMathWalker) [inline src/models/comment.rs]: With this structure, it's still possible for a user to ignore your `new` constructor and create a `Comment` using an empty string:
```rust
Comment { comment: "" }
```
It would be better to either make `comment` private or use a wrapper type around `String`, as we did for `Title` in `Ticket`.

File: src/store.rs [hunk 2]
```diff
@@ -72,6 +73,11 @@ impl TicketStore {
     pub fn update_ticket_status(&mut self, id: TicketId, status: Status) -> Option<()> {
         self.data.get_mut(&id).map(|t| t.status = status)
     }
+
+    pub fn add_comment_to_ticket(&mut self, id: TicketId, comment: String) -> Option<()> {
```
Comment (LukeMathWalker) [inline src/store.rs]: This function will panic if you pass in an empty string as a comment. 
I'd suggest to push validation at the edges: require `comment` to be of type `Comment`, hence you can rely on the invariant that it's going to be non-empty and you don't have to perform any validation.

File: src/store.rs [hunk 3]
```diff
@@ -281,4 +287,42 @@ mod tests {
...
+    fn add_comment_to_ticket() {
+        //arrange
+        let mut ticket_store = TicketStore::new();
+        let ticket = generate_and_persist_ticket(&mut ticket_store);
+        //act
+        let result = ticket_store.add_comment_to_ticket(ticket.id, "Test Comment".to_string());
+        //assert
+        assert!(result.is_some());
+        let ticket = ticket_store.get(ticket.id).unwrap();
+        assert_eq!(ticket.comments.len(), 1);
```
Comment (LukeMathWalker) [inline src/store.rs]: You can also check that it's the same comment you inserted :D

File: src/store.rs [hunk 4]
```diff
@@ -281,4 +287,42 @@ mod tests {
...
+
+        //act
+        let result = ticket_store.add_comment_to_ticket(faker.fake(), faker.fake());
+
+        //assert
+        assert!(result.is_none());
+    }
+
+    #[test]
+    #[should_panic(expected = "Comment cannot be empty")]
```
Comment (LukeMathWalker) [inline src/store.rs]: I didn't know you could assert on the panic message, neat 💯

**Comment by LukeMathWalker** [comment]
I am okay with the overall structure 👍 
Added some comments here and there to streamline the validation bit.

File: src/models/comment.rs (hunk 1 above)
Comment (james2509) [reply src/models/comment.rs]: Cool. I've removed the pub modifier. Not sure why I put that in there as initial revision didn't have it.

File: src/store.rs (hunk 2 above)
Comment (james2509) [reply src/store.rs]: I think I'm on the same page. I've updated add_comment_to_ticket to take a Comment instead of string.

File: src/store.rs (hunk 4 above)
Comment (james2509) [reply src/store.rs]: 👍 This test has been removed now as it no longer needs to test that the  add_comment_to_ticket function is panicking as that now takes a Comment structure.

File: src/store.rs (hunk 3 above)
Comment (james2509) [reply src/store.rs]: I've updated this although I think after I do more research I could refactor this to be a bit less verbose.

File: src/store.rs [hunk 5]
```diff
@@ -281,4 +286,36 @@ mod tests {
...
+        let comment = Comment::new("Test Comment".to_string()).unwrap();
+        let expected = Comment::new("Test Comment".to_string()).unwrap();
+
+        //act
+        let result = ticket_store.add_comment_to_ticket(ticket.id, comment);
+        //assert
+        assert!(result.is_some());
+        let ticket = ticket_store.get(ticket.id).unwrap();
+        assert_eq!(ticket.comments.len(), 1);
+        assert_eq!(ticket.comments[0], expected);
```
Comment (LukeMathWalker) [inline src/store.rs]: You could rephrase this as `assert_eq!(ticket.comments, vec![comment])`

**Comment by LukeMathWalker** [comment]
Merged the CLI and the persistence into master - can you merge it back into this branch and resolve the small conflicts?

File: src/store.rs (hunk 5 above)
Comment (james2509) [reply src/store.rs]: I have updated to rephrase as suggested. I couldn't do it exactly like that as comment does not - and from my understanding cannot - implement Copy. I have cloned comment into and expected comment and rephrased to `assert_eq!(ticket.comments, vec![expected_comment]`

Comment (james2509) [reply src/store.rs]: I have also merged Master into this branch and resolved issues. I've added Serialization to comment too. If you're good with my test update I'll merge into Master.

**Review by LukeMathWalker — APPROVED** [review]
